logging/exp/logger.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package exp is an experimental version of logging using slog.
+package exp
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/exp/slog"
+)
+
+// contextKey is a private string type to prevent collisions in the context map.
+type contextKey string
+
+// loggerKey points to the value in the context where the logger is stored.
+const loggerKey = contextKey("logger")
+
+var (
+	// defaultLogger is the default logger. It is initialized once per package
+	// include upon calling DefaultLogger.
+	defaultLogger     *slog.Logger
+	defaultLoggerOnce sync.Once
+)
+
+// NewFromEnv creates a new logger from env vars.
+// Set envPrefix+"LOG_LEVEL" to overwrite log level. Default log level is warning.
+// Set envPrefix+"LOG_MODE" to overwrite log mode. Default log mode is production.
+func NewFromEnv(envPrefix string) *slog.Logger {
+	logMode := strings.ToLower(strings.TrimSpace(os.Getenv(envPrefix + "LOG_MODE")))
+	devMode := logMode == "development"
+
+	level, err := LookupLevel(os.Getenv(envPrefix + "LOG_LEVEL"))
+	if err != nil {
+		mini := slog.New(slog.NewJSONHandler(os.Stderr))
+		mini.Warn("invalid log level, defaulting to info", "error", err)
+		level = LevelInfo
+	}
+
+	opts := &slog.HandlerOptions{
+		AddSource:   devMode,
+		Level:       level,
+		ReplaceAttr: cloudLoggingAttrsEncoder(),
+	}
+
+	var handler slog.Handler
+	if devMode {
+		handler = opts.NewTextHandler(os.Stdout)
+	} else {
+		handler = opts.NewJSONHandler(os.Stdout)
+	}
+	return slog.New(handler)
+}
+
+// Default creates a default logger.
+func Default() *slog.Logger {
+	defaultLoggerOnce.Do(func() {
+		defaultLogger = NewFromEnv("")
+	})
+	return defaultLogger
+}
+
+// WithLogger creates a new context with the provided logger attached.
+func WithLogger(ctx context.Context, logger *slog.Logger) context.Context {
+	return context.WithValue(ctx, loggerKey, logger)
+}
+
+// FromContext returns the logger stored in the context. If no such logger
+// exists, a default logger is returned.
+func FromContext(ctx context.Context) *slog.Logger {
+	if logger, ok := ctx.Value(loggerKey).(*slog.Logger); ok {
+		return logger
+	}
+	return Default()
+}
```

Comment:
Yea, slog has package functions for this, but we want to control the default value and explicitly want something different from the globally modified value.

Reply:
Removed in the latest proposal! So indeed you should add back.